openverse_catalog/dags/providers/provider_csv_load_scripts/inaturalist/01_photos.sql

```diff
@@ -0,0 +1,45 @@
+/*
+-------------------------------------------------------------------------------
+PHOTOS
+-------------------------------------------------------------------------------
+--  not using an FK constraint on observer_id to save load time, but the
+    5/30/2022 dataset does have complete observer ids
+--  not using an FK constraint on observation_uuid to save load time, but the
+    5/30/2022 dataset does have complete observer ids
+--  photo_id is not unique. There are ~130,000 photo_ids that appear more than
+    once, maybe because an earlier version of the photo was deleted (unclear),
+    but for now assuming that will be taken care of later in the processing.
+
+Taking DDL from
+https://github.com/inaturalist/inaturalist-open-data/blob/main/Metadata/structure.sql
+*/
+
+DROP TABLE IF EXISTS inaturalist.photos CASCADE;
+COMMIT;
+
+CREATE TABLE inaturalist.photos (
+    photo_uuid uuid NOT NULL,
+    photo_id integer NOT NULL,
+    observation_uuid uuid NOT NULL,
+    observer_id integer,
+    extension character varying(5),
+    license character varying(255),
```

Comment:
As far as I understand it, Postgres text-like columns are usually better off as `text` rather than Postgres's varchar implementation. Is there a reason to use `character` here instead of `text`?

`text` appears to have some subtle advantages over the constrained `character` types.

From https://www.postgresql.org/docs/9.1/datatype-character.html:

> Tip: There is no performance difference among these three types [`text`, `character varying(n)`, and `character(n)`], apart from increased storage space when using the blank-padded type, and a few extra CPU cycles to check the length when storing into a length-constrained column. While character(n) has performance advantages in some other database systems, there is no such advantage in PostgreSQL; in fact character(n) is usually the slowest of the three because of its additional storage costs. In most situations text or character varying should be used instead.

Reply:
TIL `text` is as fast as `varchar` in postgres. I'm remembering the anti-text conversations I've had in the past, and that the issue was really about the time needed to retrieve very long strings stored outside of the normal tablespace. ("Very long values are also stored in background tables so that they do not interfere with rapid access to shorter column values.") I guess we can trust that iNaturalist did the check on their end, since these fields are from their DDL, so we shouldn't find any very long strings in the text fields and it should be a non-issue. You think just ditch all the varchar for text @sarayourfriend ? @AetherUnbound any thoughts?

Reply:
I think text is actually slightly faster, because it doesn't need to do a length check before saving it. It's also smaller because it doesn't pad the string. Given the number of rows here, it could be a significant amount of "padded" characters taking up unused space.

As far as I understand it, in Postgres text is the way to go unless you truly have a length constraint you want to enforce. My vote is to ditch the varchar and go for text. It's also one more variable we don't have to think about if avoidable (length).

Reply:
Hmmm... right, so these are varchar, not char, which means that the padding is not an issue. And, I guess, right, I was thinking that we *don't* have to worry about the length, *because* inaturalist would have already worried about it for us. Because their schema uses varchar, so I agree, why check the length *again* on our end. 

My only hesitation is that, at least from my understanding, it's not really a free-for-all with text in postgres. You can have text up to I think a GB or maybe more, but if you want to access the very long values, it will be slower for those specific values because postgres doesn't store them with the rest of the normal table data. (And that's on purpose, to give the flexibility for very long values while still making it faster to read the shorter text values.) Does this make sense?  I'm really just looking for a "yes, trust inaturalist to keep it short or medium length, and don't worry about an extreme worst case scenario where somehow the code is trying to load the whole taxa table into each file type field."

Reply:
> so these are varchar, not char, which means that the padding is not an issue

Oops! Thanks for pointing out this difference. Sorry I missed it.

I think we can safely trust that things like the license description and extension are not going to be that long. If they _are_ then I'd expect us to explicitly truncate them in the code. Though I suppose in that case if something suddenly became unexpectedly long then the character limit on the column would cause the DAG to fail and alert us that it needed to be changed to accommodate the change in data.

I suppose it doesn't hurt to guard it from being absurdly long. If we ignore the performance question (you've convinced me that it's not worth worrying about it here), then my only hesitation is that the lengths for some of these are potentially meaningless/arbitrary which can be confusing for someone down the line trying to understand intent. Are license fields only allowed to be 255 characters long because of some standard? Is this a hard enforcement? If one day they turn out to validly be 256, is this an acceptable change that we should make to the column length to accommodate it?

Those kinds of questions worry me from the perspective of long term maintainability and intelligibility of the code.

I personally feel fine trusting that iNaturalist is exporting sensible data. If we don't trust them to export sensible data then are there other assurances we need to make elsewhere?

Reply:
Great points! I think in general, I would agree that `text` would be a better field to use over `character varying` when a hard limit is not required. However, my only hesitation here is that this is the schema pulled directly from the iNaturalist open data repo:
https://github.com/inaturalist/inaturalist-open-data/blob/5b7804f9f6d073c5a7ab52b8cb992e0afb1f0f7d/Metadata/structure.sql#L17-L18

That means (as I understand it) that it'd be impossible to receive larger data from those columns because that's the length of the column at the source, so we could use `text` here without concern. On the other hand, it introduces drift from the upstream schema which may be difficult to parse out if future changes upstream occur.

I think it'd be best to stick to the upstream schema in this case and use `character varying`. If we chose to go with `text` we should add comments where we do change those values as notes for ourselves in the future that those were columns we opted to change and aren't related to any upstream schema changes.

Reply:
Ah! If they match the upstream schema 1-to-1 then nothing that somewhere would be helpful as well, should the schema change in the future. If that's already noted then please ignore, I've just missed it.

Reply:
Yeah, I could go either way. Reading [this](https://dba.stackexchange.com/questions/20974/should-i-add-an-arbitrary-length-limit-to-varchar-columns), and in revisiting some of the json stuff,  I will say that I'm really getting in touch with the limitations of my early sql training. :) The conversation I just linked also takes into consideration whether this will be the kind of table where changing the definition in future would be a pain, but I think since these tables are really just temporary holding spots for the original tsv from inaturalist (as opposed to being designed for ongoing catalog use) it's really six of one half dozen of the other. But if we get to a point where, for example, we start doing incremental updates and these tables really need to stick around in between... BTW, should there be a step to drop them at the end of the dag?

Reply:
I'm not sure if this link will work properly, but the source is referenced at the top of each SQL file! https://github.com/WordPress/openverse-catalog/pull/549/files/212f91d50257b434eeef3e3eb896636d0a8a6f99#diff-f74331ca3b543816cf39fa02b9be40a33cdfe59f6f938e0e094fb62cc48b35c6R13-R14

And oh shoot, I totally forgot about that step 😅 It wouldn't be a bad idea to have those too. They'll be dropped on successive runs but we don't need to keep them around, really. You could tack that onto the end of the `pull_data` task within the provider ingester class functions, or we could potentially add a `post_ingestion_tasks` list as well. I think the former might be easiest in this case.

Reply:
Hmmm... I have an idea of how to create post-ingestion tasks based on pre-ingestion tasks, but there are dag parsing tests that I'm not sure how to navigate. It's a very simple sql -- `DROP SCHEMA IF EXISTS inaturalist CASCADE` -- but I'm not sure how or where to tack it on to the end of `pull_data`. ???